src/oplog/log.js

```diff
@@ -222,54 +221,86 @@ const Log = async (identity, { logId, logHeads, access, entryStorage, headsStora
    *
    * @example
    *
-   * await log.join(entry)
+   * await log.joinEntry(entry)
    *
    * @memberof module:Log~Log
    * @instance
    */
   const joinEntry = async (entry) => {
-    const { hash } = entry
-    // Check if the entry is already in the log and return early if it is
-    const isAlreadyInTheLog = await has(hash)
+    /* 1. Check if the entry is already in the log and return early if it is */
+    const isAlreadyInTheLog = await has(entry.hash)
     if (isAlreadyInTheLog) {
       return false
-    } else {
-      // Check that the entry is not an entry that hasn't been indexed
-      const it = traverse(await heads(), (e) => e.next.includes(hash) || entry.next.includes(e.hash))
-      for await (const e of it) {
-        if (e.next.includes(hash)) {
-          await _index.put(hash, true)
-          return false
-        }
-      }
     }
-    // Check that the Entry belongs to this Log
-    if (entry.id !== id) {
-      throw new Error(`Entry's id (${entry.id}) doesn't match the log's id (${id}).`)
-    }
-    // Verify if entry is allowed to be added to the log
-    const canAppend = await access.canAppend(entry)
-    if (!canAppend) {
-      throw new Error(`Could not append entry:\nKey "${entry.identity}" is not allowed to write to the log`)
+
+    const verifyEntry = async (entry) => {
+      // Check that the Entry belongs to this Log
+      if (entry.id !== id) {
+        throw new Error(`Entry's id (${entry.id}) doesn't match the log's id (${id}).`)
+      }
+      // Verify if entry is allowed to be added to the log
+      const canAppend = await access.canAppend(entry)
+      if (!canAppend) {
+        throw new Error(`Could not append entry:\nKey "${entry.identity}" is not allowed to write to the log`)
+      }
+      // Verify signature for the entry
+      const isValid = await Entry.verify(identity, entry)
+      if (!isValid) {
+        throw new Error(`Could not validate signature for entry "${entry.hash}"`)
+      }
     }
-    // Verify signature for the entry
-    const isValid = await Entry.verify(identity, entry)
-    if (!isValid) {
-      throw new Error(`Could not validate signature for entry "${hash}"`)
+
+    /* 2. Verify the entry */
+    await verifyEntry(entry)
+
+    /* 3. Find missing entries and connections (=path in the DAG) to the current heads */
+    const headsHashes = (await heads()).map(e => e.hash)
+    const hashesToAdd = new Set([entry.hash])
+    const hashesToGet = new Set([...entry.next, ...entry.refs])
+    const connectedHeads = new Set()
+
+    const traverseAndVerify = async () => {
+      const getEntries = Array.from(hashesToGet.values()).filter(has).map(get)
+      const entries = await Promise.all(getEntries)
+
+      for (const e of entries) {
+        hashesToGet.delete(e.hash)
+
+        await verifyEntry(e)
```

Comment:
I wonder if there is a case where you'd like to skip a single invalid entry? It looks like an exception in `verifyEntry` would cause `joinEntry` to stop. But maybe there are cases where you'd like to continue with other `nexts`.

Reply:
There should not be any invalid entries in any part of the oplog DAG. If there were, it’d be a malformed DAG (malicious or otherwise).